src/security/KeyData.cc

```diff
@@ -42,6 +43,40 @@ Security::KeyData::loadX509CertFromFile()
         return false;
     }
 
+    try {
+        // selected bundled certificates in sending order: wireCerts = cert + chain
+        CertList wireCerts(1, cert);
+
+        while (const auto bundledCert = Ssl::ReadOptionalCertificate(bio)) {
+            assert(!wireCerts.empty()); // this->cert is there (at least)
+
+            // We cannot chain any certificate after a self-signed certificate. This
+            // check also protects the IssuedBy() check below from adding duplicated
+            // (i.e. listed multiple times in the bundle) self-signed certificates.
+            if (SelfSigned(*wireCerts.back())) {
+                debugs(83, DBG_PARSE_NOTE(2), "WARNING: Ignoring certificate after a self-signed one: " << *bundledCert);
+                continue; // ... but keep going to report all ignored certificates
+            }
+
+            // add the bundled certificate if it extends the chain further
+            if (IssuedBy(*wireCerts.back(), *bundledCert)) {
+                debugs(83, DBG_PARSE_NOTE(3), "Adding issuer CA: " << *bundledCert);
+                // OpenSSL API requires that we order certificates such that the
+                // chain can be appended directly into the on-wire traffic.
+                chain.emplace_back(bundledCert);
+                wireCerts.emplace_back(bundledCert);
+                continue;
+            }
+
+            debugs(83, DBG_PARSE_NOTE(2), "WARNING: Ignoring certificate that does not extend the chain: " << *bundledCert);
+        }
+    }
+    catch (...) {
+        // TODO: Reject configs with malformed intermediate certs instead.
+        debugs(83, DBG_IMPORTANT, "ERROR: Failure while loading intermediate certificate(s) from '" << certFile << "':" <<
+               Debug::Extra << "problem: " << CurrentException);
+    }
```

Comment:
All these try/catch blocks, old and moved, are ugly. They are needed for a "minimal" PR that avoids changing how various certificate errors are handled. They will be gone in subsequent PRs that focus on error handling. Meanwhile, I recommend ignoring them when interpreting PR code.